Make Stream.withFilter.{map,flatMap} run in constant stack space

Blaisorblade opened the pull request:
The included test currently fails because `map` and `flatMap` do not
run in constant stack space on a stream returned by `Stream.withFilter`,
as I reported here:
https://groups.google.com/d/msg/scala-language/WqJR38REXnk/saaSiDdmyqoJ

Fix the problem and add a simple testcase.

Note that the stack space consumed when producing an element of this stream is
proportional to the number of elements failing the test before the next
success. The stack space consumed to produce the stream itself is the
space needed to produce the first element, that is, is proportional to
the number of failures before the first success.

Review by @axel22.

Blaisorblade committed:
Make Stream.withFilter.{map,flatMap} run in constant stack space

The included test currently fails because `map` and `flatMap` do not
run in constant stack space on a stream returned by `Stream.withFilter`,
as I reported here:
https://groups.google.com/d/msg/scala-language/WqJR38REXnk/saaSiDdmyqoJ

Fix the problem and add a simple testcase.

Note that the stack space consumed when producing an element of this stream is
proportional to the number of elements failing the test before the next
success. The stack space consumed to produce the stream itself is the
space needed to produce the first element, that is, is proportional to
the number of failures before the first success.

Blaisorblade commented on test/files/run/stream-stack-overflow-filter-map.scala:
```diff
@@ -0,0 +1,14 @@
+  assert(resMap1.isEmpty)
+  assert(resFMap1.isEmpty)
+  println(resMap1)
+  println(resFMap1)
```
On second thought, I could remove these lines and have no output at all. I'll push a new commit if this is preferred.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/249/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/956/

Blaisorblade commented on test/files/run/stream-stack-overflow-filter-map.scala:
Here I meant "with the previous implementation".

Blaisorblade committed:
Remove commented out code

Blaisorblade committed:
Cleanup testcase

No need to check the output - checking programmatically that the
produced streams are empty is enough.

Blaisorblade commented on src/library/scala/collection/immutable/Stream.scala:
```diff
@@ -479,22 +479,57 @@ self =>
         else tailMap
+        //XXX Alternative: what about having a Stream.step constructor?
       )
+      */
```
Commenting out code instead of removing it was not my intention. Will fix.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/956/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/249/

Blaisorblade commented:
PLS REBUILD ALL

Blaisorblade commented:
Note: the test results are, strictly speaking, not valid any more, even though the changes are minor. I would have deleted the comments, as required by the pull request policy, but I don't have permissions for that.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/255/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/962/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/255/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/962/

axel22 commented:
I believe this would benefit from additional correctness test cases such as calling `withFilter` and `map`/`flatMap` on the:
- empty stream
- one-element stream
- two-element stream

In all cases, the predicate for `withFilter` should a) always return true, b) always return false, c) `_ % 2 == 0`, d) `_ % 3 == 0`.
In all cases, the result should be compared against the expected result to see if the correct stream elements are returned.

After that, LGTM.

Blaisorblade committed:
Improve test for Stream.withFilter.{map,flatMap}

Test a wider range of functionality.

Blaisorblade committed:
Also check that Stream.toSeq gives the right result.

Blaisorblade commented:
I wrote the test case you asked for - however, the "expected" result is computed by converting the stream to a Seq (and checking that the result is right) and using filter, map and flatMap on Seq. I thus assume that those functions on Seq are are tested independently. If that's a problem I could list instead the results explicitly, even though that would make additional testcases harder to add. Or I could additionally print the results (and have a corresponding .check file).

Blaisorblade commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/266/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/973/

axel22 commented:
LGTM

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/973/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/266/
